[APM] Anomaly detection integration with transaction duration chart

ogupte opened the pull request:
Closes https://github.com/elastic/kibana/issues/70859
Reintroduces the previous anomaly detection ML integration back into the transaction duration chart in the service details screen. Support the latest APM anoamly detection by environment jobs.

![transaction-duration-chart-anomaly-detection-0](https://user-images.githubusercontent.com/1967266/86908380-48c53800-c0cb-11ea-807d-fa37e4414b0c.gif)

mistic committed:
Revert "[i18n] integrate new translations + new i18n check #70193 (#70423)"

This reverts commit 2212beba6847051bd42f759ec2530d7c9521e64d.

ogupte committed:
Reintroduces the previous anomaly detection ML integration back into the
transaction duration chart in the service details screen. Support the
latest APM anoamly detection by environment jobs.

sorenlouv commented on x-pack/plugins/apm/public/components/app/TransactionOverview/index.tsx:
```diff
@@ -130,8 +130,7 @@ export function TransactionOverview() {
             <TransactionCharts
-              // TODO [APM ML] set hasMLJob prop when ML integration is reintroduced:
-              hasMLJob={false}
+              hasMLJob={transactionCharts.mlJobId !== undefined}
```
This is not needed since you are already passing in `transactionCharts`...
```suggestion

```

sorenlouv commented on x-pack/plugins/apm/public/components/shared/charts/TransactionCharts/index.tsx:
```diff
@@ -96,18 +96,19 @@ export class TransactionCharts extends Component<TransactionChartProps> {
+      hasMLJob,
+      charts: { mlJobId },
+    } = this.props;
+    if (!hasValidMlLicense || !hasMLJob || !mlJobId) {
```
What's the difference between `!hasMLJob` and `!mlJobId` ?

ogupte replied:
There's really no difference other than types, but the the bang coerces either type to a boolean, so i can safely remove `hasMLJob`.

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/fetcher.ts:
```diff
@@ -0,0 +1,105 @@
+            { term: { job_id: jobId } },
+            { exists: { field: 'bucket_span' } },
+            {
+              term: {
```
Can you compact these?

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/fetcher.ts:
```diff
@@ -0,0 +1,105 @@
+  try {
+    const response = await ml.mlSystem.mlAnomalySearch(params);
+    return response;
+  } catch (err) {
+    const isHttpError = 'statusCode' in err;
+    if (isHttpError) {
+      return;
+    }
+    throw err;
+  }
```
I don't think we should re-throw since this is an external integration, and if it throws none of the transaction charts will show up

```suggestion
  } catch (err) {
    const isHttpError = 'statusCode' in err;
    if (isHttpError) {
      logger.info(`Status code "${err.statusCode}" while retrieving ML anomalies for APM`)
      return;
    }
  
    logger.error('An error occurred while retrieving ML anomalies for APM')
    logger.error(err)
  }
```

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/index.ts:
```diff
@@ -55,6 +66,40 @@ export async function getAnomalySeries({
+    jobId,
+  });
+
+  return esResponse
+    ? anomalySeriesTransform(
+        esResponse,
+        mlBucketSize,
+        bucketSize,
+        timeSeriesDates,
+        jobId
+      )
+    : undefined;
```
An `if` is simpler than a ternary in this case:

```suggestion
  if(esResponse && mlBucketSize) {
    return anomalySeriesTransform(
        esResponse,
        mlBucketSize,
        bucketSize,
        timeSeriesDates,
        jobId
      )
  }
```

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/get_ml_bucket_size.ts:
```diff
@@ -0,0 +1,65 @@
+  } catch (err) {
+    const isHttpError = 'statusCode' in err;
+    if (isHttpError) {
+      return 0;
```
returning 0 seems wrong since it could result in downstream errors if we actually try to use a bucketsize of 0

```suggestion
      return;
```

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/get_ml_bucket_size.ts:
```diff
@@ -0,0 +1,65 @@
+    if (isHttpError) {
+      return 0;
+    }
+    throw err;
```
Let's not throw
```suggestion
    logger.error(...)
```

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/index.ts:
```diff
@@ -39,9 +43,16 @@ export async function getAnomalySeries({
+    // filter out known uiFilters like service.environment & service.name
+    const unknownFilters = setup.uiFiltersES.filter(
+      (uiFilter) =>
+        !uiFilter.term?.[SERVICE_ENVIRONMENT] && !uiFilter.terms?.[SERVICE_NAME]
```
I wouldn't think service name was part of ui filters?

ogupte replied:
I think it gets included here: https://github.com/elastic/kibana/blob/master/x-pack/plugins/apm/public/components/app/TransactionOverview/index.tsx#L101

lizozom committed:
Allow creating filters from fields with null values in discover (#70936)

* Fix bug #7189

* typo

* Test adjustments

* wait for load complete

* Fine tune test

* Update src/plugins/data/public/query/filter_manager/lib/generate_filters.ts

Co-authored-by: Lukas Olson <olson.lukas@gmail.com>

* Fix filtering by an array of nulls
Allow filtering by a non existing field in the doc
simplify flatten hit logic

Co-authored-by: Lukas Olson <olson.lukas@gmail.com>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

dej611 committed:
Improve failure reporting output for functional tests (#70954)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

thomheymann committed:
Clone role button going to edit page (#70768)

* Clone role button going to edit page

* Added unit tests

* Fixed types

sorenlouv commented on x-pack/plugins/apm/public/selectors/chartSelectors.ts:
```diff
@@ -62,6 +63,7 @@ export function getTransactionCharts(
   return {
     tpmSeries,
     responseTimeSeries,
+    mlJobId: anomalyTimeseries?.jobId,
```
This doesn't seem like the best place to put `mlJobId` but okay for now.
Would be good to go over chartSelectors and see if we can move it to the backend or something like it.

sorenlouv commented on x-pack/plugins/apm/server/lib/transactions/charts/get_anomaly_data/get_ml_bucket_size.ts:
```diff
@@ -0,0 +1,65 @@
+}: IOptions): Promise<number> {
+  const { ml, start, end } = setup;
+  if (!ml) {
+    return 0;
```
```suggestion
    return;
```

mistic committed:
Merge remote-tracking branch 'upstream/master'

mistic committed:
skip flaky suite (#71220)

flash1293 committed:
mitigate kql bug (#70712)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

rudolf committed:
Disable fields added from unknown saved object types (#70951)

* Allow disabled: false SO field mappings

* Disable fields for unknown SO types

* Update everyone else's docs ;)

* Address review comments

* Add unit tests for disableUnknownTypeMappingFields()

ogupte committed:
[APM] Service maps anomaly detection integration by environment (#70932)

* Closes #69480 & #70419.
- Adds anomaly detection integration to service maps backed by apm ML jobs per environment
- Loads transaction stats and anomalies for each transaction types
- Renders a selector in the popop to choose a transaction type to view stats

* - implements original anomaly detection integration design for service maps popover
- only aggregates transaction KPIs and anomaly scores for transaction.type = "request" or "page-load"
- supports environment filter 'All' option to display data from all APM anomaly detection jobs
- handle case where popover metrics don't exist for services outside the current environment filter

* fixes some CI errors

* Simplified messaging for service popop with not data in the current environment

* PR feedback, renamed max anomalies -> service anomalies including the file name

* - defines custom_settings.job_tags.apm_ml_version in ML job creation,
  then filters for it when returing valid APM ML jobs

* changes shape of of service anomalies from an array to a object keyed by serviceName

* removes the url encoding from ML job link href to how it was previously.

* PR feedback

* Popover no data state simplified:
- renders the "no data" message as plain text instead of in a callout
- hides the 'Anomaly detection' section if there is not anomaly data.

* Fixes filtering bug when user selects 'Environment: Not defined'. Now
filters properly by filtering for docs where service.environment does
not exist

* filters jobs fetched in the settings page by `job.custom_settings.job_tags.apm_ml_version`

* Fixed bad import from last upstream merge

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

kertal committed:
[Discover] Improve saved queries reload functional test (#70957)

kertal committed:
[Discover] Improve Discover histogram time range tests (#71056)

* Improve Discover histogram tests

* Use includes function to find a string

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

mistic committed:
skip flaky suite (#71216)

nchaulet committed:
[Ingest Manager] Fix and improve agent status (#71009)

Commit:
Unskip data table non timebased test (#71049)

* Unskip data table non timebased test

* Fix test to use new label

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

alisonelizabeth committed:
[Index Management] Fix broken app links (#71007)

andrewvc committed:
[Uptime] Remove Scripted Metric Query (#69229)

Resolves #68096 , removing unnecessary scripted metric query from overview page and unifying the Check and Ping types.

This simplifies the types in a number of ways, and reduces the total quantity of code to execute the queries for the overview page. It also fixes the Tls and related types which were inconsistent and presented a problem here since they are used by this JS. There are now three stages where before there were four:

    Find potential matches: where we determine which monitor IDs are eligible for the overview page
    Refine potential matches: where we determine which ones actually match and return the summary documents for each location to build the MonitorSummary object
    Get monitor histograms: where we calculate the histograms for each monitor. In the future we might make this a separate API call.

This improves the overall code structure, and leaves the test coverage about the same depending on how you look at it. I think we can do more to improve the quality of code / tests here, but this seemed like a good initial place to draw the line for now.

In perfunctory testing on our internal observability clusters I saw perf improve from 2.5s to 1.1s on the Uptime homepage with no filters. So, it looks like there are potentially perf improvements (no real benchmarking was done).

Previously, this returned all pings from the latest check group. This was not actually used anywhere, only the summary pings are required for the current UI, so we now only return those from the list API as this saves a query.

rudolf committed:
Test that savedObjectsTable exists AND isn't loading (#71207)

* Wait until savedObjectsTable exists AND isn't loading

* Use :not css selector

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

qn895 committed:
[ML] Add switch to enable model plot annotations independently (#70678)

walterra committed:
[ML] string_utils cleanup. (#71092)

- Converts string_utils to TypeScript.
- Removes sortByKey() from string_utils, we no longer make use of it.
- Fixes #69499, stringMatch() was defined twice, now moved to string_utils.
- Fixes #69498, OMIT_FIELDS was defined twice, now moved to common/constants/field_types.ts.

tylersmalley committed:
Updates translations

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

YulNaumenko committed:
Changed "Elastic" to "Elastic Alerts" in PagerDuty action doc. (#71184)

mistic committed:
skip flaky suite (#65949)

nnamdifrankie committed:
[SecuritySolution - Endpoint]: update payload for metadata and policy response tests (#71084)

[SecuritySolution - Endpoint]: update payload for metadata and policy response tests

mistic committed:
skip flaky suite (#69849)

KOTungseth committed:
[DOCS] Image cleanup (#71258)

* [DOCS] Image clean up

* Second half of images folder

* Fixes errors

jloleysens committed:
[Index Management] Fix get data streams api integration (#71067)

* fix get all data streams api integration test

* fix and skip api integration with ES for fetching one data stream

* update types

* fix types issue

* unskip tests

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

ThomThomson committed:
[Example] Embeddable by Reference and Value (#68719)

Added an attribute service to embeddable start contract which provides a higher level abstraction for embeddables that can be by reference OR by value. Added an example that uses this service.

efreeti committed:
Fixing the consistency of util bar in endpoints and policies pages. (#71133)

tsullivan committed:
[Reporting] remove async execution from csv_from_savedobject  (#71031)

* [Reporting] remove async execution from csv_from_savedobject

This simplifies the csv_from_savedobject logic by removing the async
hook. This was added as premature optimization in the initial PR that
added the Download CSV button to the dashboards.

* copy out export type ts changes

* remove routes

* fix i18n

patrykkopycinski committed:
[Security Solution] Manage timeline templates user flow (#67837)

cauemarcondes committed:
[Observability] Fetch news feed (#71212)

* adding new feed

* adding log when an error happens

* fixing translations

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

alisonelizabeth committed:
[Console] Fix example in help flyout (#71188)

alvarezmelissa87 committed:
get rid of console err in management. disable links in menu if needed (#71118)

madirey committed:
Preserve compressed artifacts when updating manifest (#71196)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

ThomThomson committed:
Dashboard add or update panel (#71130)

Added a standard method for adding or replacing a panel on a dashboard.

FrankHassanabad committed:
[SIEM][Detection Engine][Lists] Adds "wait_for" to all the create, update, patch, delete endpoints

## Summary

* Adds "wait_for" to all the create, update, patch, and delete endpoints
* Ran some quick tests against import and the performance still looks acceptable
* Updates the unit tests to reflect the addition

### Checklist

- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios

yuliacech committed:
[ILM] Change "wait for snapshot" policy text field to EuiCombobox (#70627)

* [ILM] Change "Wait for snapshot policy" text field to a dropdown in Delete phase

* [ILM] Change "wait for snapshot" field to a EuiCombobox and update jest tests

* [ILM] Update jest tests to check callouts

* [ILM] Implement PR review suggestions

* Update x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/components/snapshot_policies/snapshot_policies.tsx

Co-authored-by: Adam Locke <adam.locke@elastic.co>

* Update x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/components/snapshot_policies/snapshot_policies.tsx

Co-authored-by: Adam Locke <adam.locke@elastic.co>

* Update x-pack/plugins/index_lifecycle_management/public/application/sections/edit_policy/components/snapshot_policies/snapshot_policies.tsx

Co-authored-by: Adam Locke <adam.locke@elastic.co>

* [ILM] Fix copy

* [ILM] Fix copy

* [ILM] Fix build error

* [ILM] Delete periods in callout titles

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>
Co-authored-by: Adam Locke <adam.locke@elastic.co>

cauemarcondes committed:
[Observability] illustration for landing page (#71217)

* changin illustration

* renaming files

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

neptunian committed:
Revert "[Ingest Manager ] prepend kibana asset ids with package name (#70502)" (#71271)

This reverts commit 984ea0700ee8b84e69f626792e1dd913607307c9.

shahzad31 committed:
[Uptime] Jest test adjust to use relative date (#70411)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

justinkambic committed:
[Uptime] Availability alert (#70284)

* Add new fields to monitor status alert for availability.

* Add UI code for availability threshold expression + translations/a11y fields.

* Add availability selection to alert UI.

* Disable expression popover button functionality when not enabled.

* Add select box for monitor status alert, reorganize layout.

* Add new runtime types for parsing in alert executor.

* Add new enablement field for monitor status checks.

* Add availability check query function and tests. Extract helper function from similar test file to generic file.

* Add availability checking to status check alert type.

* Change availability threshold to be number type.

* Remove clearing of fields when disabled.

* Change alert validation to require availability or status check.

* Fix threshold input functionality.

* Add tests and refine alert validation.

* Add test for new validation logic.

* Add any type temporarily.

* Delete unused code, fix types.

* Add filter capabilities to availability type.

* Disable availability by default for old alerts.

* Add filtering to availability query.

* Clean up types and refresh test snapshots.

* Change threshold storage value to string. Add bucket selector agg.

* Update copy.

* Add tests and improve should check flag evaluation.

* Improve old alert detection code.

* Fix issue with status check enablement.

* Update unit tests to reflect changes to query.

* Fix types.

* Improve tests, refactor a function to clean up code.

* Remove fields from aggregate key and retrieve them from top hits instead.

* Add sort parameter to top_hits aggregation.

* Update context message of monitor status alert, and add translations for availability message.

* Modify default alert message.

* Add a comment.

* Fix outdated translations.

* Revert unknown to any to simplify validation.

* Improve readability of array manipulation for availability result description.

* Add a flex item wrapper to fix layout problem.

sebelga committed:
[Composable template] Demo and PR review fixes (#71065)

kevinlog committed:
[SECURITY_SOLUTION] adjust policy onboarding view, check for Ingest permissions (#70536)

* adjust policy onboarding view

* correct test subj

* fix tests

* re-enable tests

* add no permissions view

* adjust onbording look

* adjust text

* use ingest hook, add tests

* adjust text

* address comments

* beta badges

* fix test

* correct timeline flyout

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

jen-huang committed:
[Ingest Manager] Integrate beta messaging with Add Data (#71147)

* Add methods to register directory notices and header links in tutorials, and use registered components when rendering tutorial directory

* Add methods to register module notices components in tutorial pages, and use registered components when rendering tutorial page

* Add `moduleName` field to server tutorial schema and test fixure

* Surface `moduleName` field from built in tutorials and registered apm tutorial

* Export component types

* Add KibanaContextProvider to home plugin app render

* Move setHttpClient to ingest manager plugin setup() method; add home as optional plugin dep; register tutorial module notice

* Fix key prop warnings

* Add dismissable tutorial directory notice and corresponding ingest manager global setting field

* Add tutorial directory header link and tie it to the state of the dismissible directory notice via observable

* Put spacing inside module notice component itself

* Check if ingest manager is available in current space

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

Commit:
New Enterprise Search Kibana plugin (#66922)

* Initial App Search in Kibana plugin work

- Initializes a new platform plugin that ships out of the box w/ x-pack
- Contains a very basic front-end that shows AS engines, error states, or a Setup Guide
- Contains a very basic server that remotely calls the AS internal engines API and returns results

* Update URL casing to match Kibana best practices

- URL casing appears to be snake_casing, but kibana.json casing appears to be camelCase

* Register App Search plugin in Home Feature Catalogue

* Add custom App Search in Kibana logo

- I haven't had much success in surfacing a SVG file via a server-side endpoint/URL, but then I realized EuiIcon supports passing in a ReactElement directly. Woo!

* Fix appSearch.host config setting to be optional

- instead of crashing folks on load

* Rename plugin to Enterprise Search

- per product decision, URL should be enterprise_search/app_search and Workplace Search should also eventually live here
- reorganize folder structure in anticipation for another workplace_search plugin/codebase living alongside app_search
- rename app.tsx/main.tsx to a standard top-level index.tsx (which will contain top-level routes/state)
- rename AS->ES files/vars where applicable
- TODO: React Router

* Set up React Router URL structure

* Convert showSetupGuide action/flag to a React Router link

- remove showSetupGuide flag
- add a new shared helper component for combining EuiButton/EuiLink with React Router behavior (https://github.com/elastic/eui/blob/master/wiki/react-router.md#react-router-51)

* Implement Kibana Chrome breadcrumbs

- create shared helper (WS will presumably also want this) for generating EUI breadcrumb objects with React Router links+click behavior
- create React component that calls chrome.setBreadcrumbs on page mount
- clean up type definitions - move app-wide props to IAppSearchProps and update most pages/views to simply import it instead of calling their own definitions

* Added server unit tests (#2)

* Added unit test for server

* PR Feedback

* Refactor top-level Kibana props to a global context state

- rather them passing them around verbosely as props, the components that need them should be able to call the useContext hook

+ Remove IAppSearchProps in favor of IKibanaContext

+ Also rename `appSearchUrl` to `enterpriseSearchUrl`, since this context will contained shared/Kibana-wide values/actions useful to both AS and WS

* Added unit tests for public (#4)

* application.test.ts

* Added Unit Test for EngineOverviewHeader

* Added Unit Test for generate_breadcrumbs

* Added Unit Test for set_breadcrumb.tsx

* Added a unit test for link_events

- Also changed link_events.tsx to link_events.ts since it's just TS, no
React
- Modified letBrowserHandleEvent so it will still return a false
boolean when target is blank

* Betterize these tests

Co-Authored-By: Constance <constancecchen@users.noreply.github.com>

Co-authored-by: Constance <constancecchen@users.noreply.github.com>

* Add UI telemetry tracking to AS in Kibana (#5)

* Set up Telemetry usageCollection, savedObjects, route, & shared helper

- The Kibana UsageCollection plugin handles collecting our telemetry UI data (views, clicks, errors, etc.) and pushing it to elastic's telemetry servers
- That data is stored in incremented in Kibana's savedObjects lib/plugin (as well as mapped)
- When an end-user hits a certain view or action, the shared helper will ping the app search telemetry route which increments the savedObject store

* Update client-side views/links to new shared telemetry helper

* Write tests for new telemetry files

* Implement remaining unit tests (#7)

* Write tests for React Router+EUI helper components

* Update generate_breadcrumbs test

- add test suite for generateBreadcrumb() itself (in order to cover a missing branch)
- minor lint fixes
- remove unnecessary import from set_breadcrumbs test

* Write test for get_username util

+ update test to return a more consistent falsey value (null)

* Add test for SetupGuide

* [Refactor] Pull out various Kibana context mocks into separate files

- I'm creating a reusable useContext mock for shallow()ed enzyme components
+ add more documentation comments + examples

* Write tests for empty state components

+ test new usecontext shallow mock

* Empty state components: Add extra getUserName branch test

* Write test for app search index/routes

* Write tests for engine overview table

+ fix bonus bug

* Write Engine Overview tests

+ Update EngineOverview logic to account for issues found during tests :)
  - Move http to async/await syntax instead of promise syntax (works better with existing HttpServiceMock jest.fn()s)
  - hasValidData wasn't strict enough in type checking/object nest checking and was causing the app itself to crash (no bueno)

* Refactor EngineOverviewHeader test to use shallow + to full coverage

- missed adding this test during telemetry work
- switching to shallow and beforeAll reduces the test time from 5s to 4s!

* [Refactor] Pull out React Router history mocks into a test util helper

+ minor refactors/updates

* Add small tests to increase branch coverage

- mostly testing fallbacks or removing fallbacks in favor of strict type interface
- these are slightly obsessive so I'd also be fine ditching them if they aren't terribly valuable

* Address larger tech debt/TODOs (#8)

* Fix optional chaining TODO

- turns out my local Prettier wasn't up to date, completely my bad

* Fix constants TODO

- adds a common folder/architecture for others to use in the future

* Remove TODO for eslint-disable-line and specify lint rule being skipped

- hopefully that's OK for review, I can't think of any other way to sanely do this without re-architecting the entire file or DDoSing our API

* Add server-side logging to route dependencies

+ add basic example of error catching/logging to Telemetry route
+ [extra] refactor mockResponseFactory name to something slightly easier to read

* Move more Engines Overview API logic/logging to server-side

- handle data validation in the server-side
- wrap server-side API in a try/catch to account for fetch issues
- more correctly return 2xx/4xx statuses and more correctly deal with those responses in the front-end
- Add server info/error/debug logs (addresses TODO)
- Update tests + minor refactors/cleanup
    - remove expectResponseToBe200With helper (since we're now returning multiple response types) and instead make mockResponse var name more readable
    - one-line header auth
    - update tests with example error logs
    - update schema validation for `type` to be an enum of `indexed`/`meta` (more accurately reflecting API)

* Per telemetry team feedback, rename usageCollection telemetry mapping name to simpler 'app_search'

- since their mapping already nests under 'kibana.plugins'
- note: I left the savedObjects name with the '_telemetry' suffix, as there very well may be a use case for top-level generic 'app_search' saved objects

* Update Setup Guide installation instructions (#9)

Co-authored-by: Chris Cressman <chris@chriscressman.com>

* [Refactor] DRY out route test helper

* [Refactor] Rename public/test_utils to public/__mocks__

- to better follow/use jest setups and for .mock.ts suffixes

* Add platinum licensing check to Meta Engines table/call (#11)

* Licensing plugin setup

* Add LicensingContext setup

* Update EngineOverview to not hit meta engines API on platinum license

* Add Jest test helpers for future shallow/context use

* Update plugin to use new Kibana nav + URL update (#12)

* Update new nav categories to add Enterprise Search + update plugin to use new category

- per @johnbarrierwilson and Matt Riley, Enterprise Search should be under Kibana and above Observability
- Run `node scripts/check_published_api_changes.js --accept` since this new category affects public API

* [URL UPDATE] Change '/app/enterprise_search/app_search' to '/app/app_search'

- This needs to be done because App Search and Workplace search *have* to be registered as separate plugins to have 2 distinct nav links
- Currently Kibana doesn't support nested app names (see: https://github.com/elastic/kibana/issues/59190) but potentially will in the future

- To support this change, we need to update applications/index.tsx to NOT handle '/app/enterprise_search' level routing, but instead accept an async imported app component (e.g. AppSearch, WorkplaceSearch).
- AppSearch should now treat its router as root '/' instead of '/app_search'

- (Addl) Per Josh Dover's recommendation, switch to `<Router history={params.history}>` from `<BrowserRouter basename={params.appBasePath}>` since they're deprecating appBasePath

* Update breadcrumbs helper to account for new URLs

- Remove path for Enterprise Search breadcrumb, since '/app/enterprise_search' will not link anywhere meaningful for the foreseeable future, so the Enterprise Search root should not go anywhere
- Update App Search helper to go to root path, per new React Router setup

Test changes:
- Mock custom basepath for App Search tests
- Swap enterpriseSearchBreadcrumbs and appSearchBreadcrumbs test order (since the latter overrides the default mock)

* Add create_first_engine_button telemetry tracking to EmptyState

* Switch plugin URLs back to /app/enterprise_search/app_search

Now that https://github.com/elastic/kibana/pull/66455 has been merged in :tada:

* Add i18n formatted messages / translations (#13)

* Add i18n provider and formatted/i18n translated messages

* Update tests to account for new I18nProvider context + FormattedMessage components

- Add new mountWithContext helper that provides all contexts+providers used in top-level app
- Add new shallowWithIntl helper for shallow() components that dive into FormattedMessage

* Format i18n dates and numbers

+ update some mock tests to not throw react-intl invalid date messages

* Update EngineOverviewHeader to disable button on prop

* Address review feedback (#14)

* Fix Prettier linting issues

* Escape App Search API endpoint URLs

- per PR feedback
- querystring should automatically encodeURIComponent / escape query param strings

* Update server plugin.ts to use getStartServices() rather than storing local references from start()

- Per feedback: https://github.com/elastic/kibana/blob/master/src/core/CONVENTIONS.md#applications

- Note: savedObjects.registerType needs to be outside of getStartServices, or an error is thrown

- Side update to registerTelemetryUsageCollector to simplify args

- Update/fix tests to account for changes

* E2E testing (#6)

* Wired up basics for E2E testing

* Added version with App Search

* Updated naming

* Switched configuration around

* Added concept of 'fixtures'

* Figured out how to log in as the enterprise_search user

* Refactored to use an App Search service

* Added some real tests

* Added a README

* Cleanup

* More cleanup

* Error handling + README updatre

* Removed unnecessary files

* Apply suggestions from code review

Co-authored-by: Constance <constancecchen@users.noreply.github.com>

* Update x-pack/plugins/enterprise_search/public/applications/app_search/components/engine_overview/engine_table.tsx

Co-authored-by: Constance <constancecchen@users.noreply.github.com>

* PR feedback - updated README

* Additional lint fixes

Co-authored-by: Constance <constancecchen@users.noreply.github.com>

* Add README and CODEOWNERS (#15)

* Add plugin README and CODEOWNERS

* Fix Typescript errors (#16)

* Fix public mocks

* Fix empty states types

* Fix engine table component errors

* Fix engine overview component errors

* Fix setup guide component errors

- SetBreadcrumbs will be fixed in a separate commit

* Fix App Search index errors

* Fix engine overview header component errors

* Fix applications context index errors

* Fix kibana breadcrumb helper errors

* Fix license helper errors

* :exclamation: Refactor React Router EUI link/button helpers
- in order to fix typescript errors

- this changes the component logic significantly to a react render prop, so that the Link and Button components can have different types - however, end behavior should still remain the same

* Fix telemetry helper errors

* Minor unused var cleanup in plugin files

* Fix telemetry collector/savedobjects errors

* Fix MockRouter type errors and add IRouteDependencies export

- routes will use IRouteDependencies in the next few commits

* Fix engines route errors

* Fix telemetry route errors

* Remove any type from source code

- thanks to Scotty for the inspiration

* Add eslint rules for Enterprise Search plugin

- Add checks for type any, but only on non-test files
- Disable react-hooks/exhaustive-deps, since we're already disabling it in a few files and other plugins also have it turned off

* Cover uncovered lines in engines_table and telemetry tests

* Fixed TS warnings in E2E tests (#17)

* Feedback: Convert static CSS values to EUI variables where possible

* Feedback: Flatten nested CSS where possible

- Prefer setting CSS class overrides on individual EUI components, not on a top-level page

+ Change CSS class casing from kebab-case to camelCase to better match EUI/Kibana

+ Remove unnecessary .euiPageContentHeader margin-bottom override by changing the panelPaddingSize of euiPageContent

+ Decrease engine overview table padding on mobile

* Refactor out components shared with Workplace Search (#18)

* Move getUserName helper to shared

- in preparation for Workplace Search plugin also using this helper

* Move Setup Guide layout to a shared component

* Setup Guide: add extra props for standard/native auth links

Note: It's possible this commit may be unnecessary if we can publish shared Enterprise Search security mode docs

* Update copy per feedback from copy team

* Address various telemetry issues

- saved objects: removing indexing per #43673
- add schema and generate json per #64942
- move definitions over to collectors since saved objects is mostly empty at this point, and schema throws an error when it imports an obj instead of being defined inline
- istanbul ignore saved_objects file since it doesn't have anything meaningful to test but was affecting code coverage

* Disable plugin access if a normal user does not have access to App Search (#19)

* Set up new server security dependency and configs

* Set up access capabilities

* Set up checkAccess helper/caller

* Remove NoUserState component from the public UI

- Since this is now being handled by checkAccess / normal users should never see the plugin at all if they don't have an account/access, the component is no longer needed

* Update server routes to account for new changes

- Remove login redirect catch from routes, since the access helper should now handle that for most users by disabling the plugin (superusers will see a generic cannot connect/error screen)
- Refactor out new config values to a shared mock

* Refactor Enterprise Search http call to hit/return new internal API endpoint

+ pull out the http call to a separate library for upcoming public URL work (so that other files can call it directly as well)

* [Discussion] Increase timeout but add another warning timeout for slow servers

- per recommendation/convo with Brandon

* Register feature control

* Remove no_as_account from UI telemetry

- since we're no longer tracking that in the UI

* Address PR feedback - isSuperUser check

* Public URL support for Elastic Cloud (#21)

* Add server-side public URL route

- Per feedback from Kibana platform team, it's not possible to pass info from server/ to public/ without a HTTP call :[

* Update MockRouter for routes without any payload/params

* Add client-side helper for calling the new public URL API

+ API seems to return a URL a trailing slash, which we need to omit

* Update public/plugin.ts to check and set a public URL

- relies on this.hasCheckedPublicUrl to only make the call once per page load instead of on every page nav

* Fix failing feature control tests

- Split up scenario cases as needed
- Add plugin as an exception alongside ML & Monitoring

* Address PR feedback

- version: kibana
- copy edits
- Sass vars
- code cleanup

* Casing feedback: change all plugin registration IDs from snake_case to camelCase

- note: current remainng snake_case exceptions are telemetry keys
- file names and api endpoints are snake_case per conventions

* Misc security feedback

- remove set
- remove unnecessary capabilities registration
- telemetry namespace agnostic

* Security feedback: add warn logging to telemetry collector

see https://github.com/elastic/kibana/pull/66922#discussion_r451215760
- add if statement
- pass log dependency around (this is kinda medium, should maybe refactor)
- update tests
- move test file comment to the right file (was meant for telemetry route file)

* Address feedback from Pierre

- Remove unnecessary ServerConfigType
- Remove unnecessary uiCapabilities
- Move registerTelemetryRoute / SavedObjectsServiceStart workaround
- Remove unnecessary license optional chaining

* PR feedback

Address type/typos

* Fix telemetry API call returning 415 on Chrome

- I can't even?? I swear charset=utf-8 fixed the same error a few weeks ago

* Fix failing tests

* Update Enterprise Search functional tests (without host) to run on CI

- Fix incorrect navigateToApp slug (hadn't realized this was a URL, not an ID)
- Update without_host_configured tests to run without API key
- Update README

* Address PR feedback from Pierre

- remove unnecessary authz?
- remove unnecessary content-type json headers
- add loggingSystemMock.collect(mockLogger).error assertion
- reconstrcut new MockRouter on beforeEach for better sandboxing
- fix incorrect describe()s -should be it()
- pull out reusable mockDependencies helper (renamed/extended from mockConfig) for tests that don't particularly use config/log but still want to pass type definitions
- Fix comment copy

Co-authored-by: Jason Stoltzfus <jastoltz24@gmail.com>
Co-authored-by: Chris Cressman <chris@chriscressman.com>
Co-authored-by: scottybollinger <scotty.bollinger@elastic.co>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

dmlemeshko committed:
Remove IE support in functional tests (#71285)

* [ftr] remove ie support

* remove ie integration tests config

lcawl committed:
[DOCS] Clarify trial subscription levels (#70636)

bkimmel committed:
[Security_Solution][Resolver]Add beta badge to Resolver panel (#71183)

* Add beta badge to Resolver panel

legrego committed:
Support multiple features declaring same properties (#71106)

Co-authored-by: Joe Portner <5295965+jportner@users.noreply.github.com>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

jen-huang committed:
[Ingest Manager] Fix limited packages incorrect response (#71292)

* Fix limited packages incorrect response

* Add test to make sure we don't break this again :D

dmlemeshko committed:
[ftr] use typed chromeOptions object, adding TEST_BROWSER_BINARY_PATH (#71279)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

skh committed:
[Ingest Manager] Add schema to usageCollector. (#71219)

* Add schema to Ingest Manager usageCollector.

* Use long for counters.

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

spalger committed:
fix visual regression job (#70999)

spalger committed:
send slack notifications on visual baseline failures

dmlemeshko committed:
[services/remote/webdriver] fix eslint error (#71346)

Commit:
Deduplication of entries and items before sending to endpoint (#71297)

* Deduplication of entries and items before sending to endpoint

* Renaming

spalger committed:
jenkins_xpack_saved_objects_field_metrics.sh expects to be run from the KIBANA_DIR in CI

cnasikas committed:
[SIEM][CASE] IBM Resilient Connector (#66385)

alvarezmelissa87 committed:
[ML] DF Analytics:  stop status polling when job stopped (#71159)

* clear interval if job stopped

* ensure analysisFieldsTable error messages up to date

* use shared constant for job state

dplumlee committed:
[Security Solution][Exceptions] - Exceptions modal pt 2 (#70886)

* makes comment updates

* adds tests

* adds back non ecs data to timeline

* comments

* fixes jest tests

* fixes typo

Commit:
[kbn/optimizer] implement "requiredBundles" property of KP plugins (#70911)

Co-authored-by: Josh Dover <me@joshdover.com>
Co-authored-by: spalger <spalger@users.noreply.github.com>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

FrankHassanabad committed:
[SIEM][Detection Engine][Lists] Adds read_privileges route for lists and list items 

## Summary

* Adds a read_privileges for the list and list items.

Run the script:
get_privileges.sh

API:

```ts
GET /api/lists/privileges

{
  "listItems": {
    "username": "yo",
    "has_all_requested": false,
    "cluster": {
      "monitor_ml": true,
      "manage_ccr": true,
      "manage_index_templates": true,
      "monitor_watcher": true,
      "monitor_transform": true,
      "read_ilm": true,
      "manage_api_key": true,
      "manage_security": true,
      "manage_own_api_key": false,
      "manage_saml": true,
      "all": true,
      "manage_ilm": true,
      "manage_ingest_pipelines": true,
      "read_ccr": true,
      "manage_rollup": true,
      "monitor": true,
      "manage_watcher": true,
      "manage": true,
      "manage_transform": true,
      "manage_token": true,
      "manage_ml": true,
      "manage_pipeline": true,
      "monitor_rollup": true,
      "transport_client": true,
      "create_snapshot": true
    },
    "index": {
      ".lists-frank-default": {
        "all": true,
        "manage_ilm": true,
        "read": true,
        "create_index": true,
        "read_cross_cluster": true,
        "index": true,
        "monitor": true,
        "delete": true,
        "manage": true,
        "delete_index": true,
        "create_doc": true,
        "view_index_metadata": true,
        "create": true,
        "manage_follow_index": true,
        "manage_leader_index": true,
        "write": true
      }
    },
    "application": {}
  },
  "lists": {
    "username": "yo",
    "has_all_requested": false,
    "cluster": {
      "monitor_ml": true,
      "manage_ccr": true,
      "manage_index_templates": true,
      "monitor_watcher": true,
      "monitor_transform": true,
      "read_ilm": true,
      "manage_api_key": true,
      "manage_security": true,
      "manage_own_api_key": false,
      "manage_saml": true,
      "all": true,
      "manage_ilm": true,
      "manage_ingest_pipelines": true,
      "read_ccr": true,
      "manage_rollup": true,
      "monitor": true,
      "manage_watcher": true,
      "manage": true,
      "manage_transform": true,
      "manage_token": true,
      "manage_ml": true,
      "manage_pipeline": true,
      "monitor_rollup": true,
      "transport_client": true,
      "create_snapshot": true
    },
    "index": {
      ".lists-frank-default": {
        "all": true,
        "manage_ilm": true,
        "read": true,
        "create_index": true,
        "read_cross_cluster": true,
        "index": true,
        "monitor": true,
        "delete": true,
        "manage": true,
        "delete_index": true,
        "create_doc": true,
        "view_index_metadata": true,
        "create": true,
        "manage_follow_index": true,
        "manage_leader_index": true,
        "write": true
      }
    },
    "application": {}
  },
  "is_authenticated": true
}
```

### Checklist

We currently have not ported over patterns for the routes so we do not have sanity checks against this or other routes and no end point tests which is why the check box is not checked below at this point in time. We are implementing those tests during the feature freeze (hopefully)

- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios

FrankHassanabad committed:
[SIEM][Detection Engine] Fixes skipped tests (#71347)

## Summary

* https://github.com/elastic/kibana/issues/69632
* Adds a retry loop in case of a network outage/issue which should increase the chances of success
* If there is still an issue after the 20th try, then it moves on and there is a high likelihood the tests will continue without issues.
* Adds console logging statements so we know if this flakiness happens again a bit more insight into why the network is behaving the way it is.
* Helps prevent the other tests from being skipped in the future due to bad networking issues. 

The errors that were coming back from the failed tests are in the `afterEach` and look to be network related or another test interfering:

```ts
1) detection engine api security and spaces enabled
01:59:54         find_statuses
01:59:54           "after each" hook for "should return a single rule status when a single rule is loaded from a find status with defaults added":
01:59:54       ResponseError: Response Error
01:59:54        at IncomingMessage.response.on (/dev/shm/workspace/kibana/node_modules/@elastic/elasticsearch/lib/Transport.js:287:25)
01:59:54        at endReadableNT (_stream_readable.js:1145:12)
01:59:54        at process._tickCallback (internal/process/next_tick.js:63:19)
01:59:54  
01:59:54               └- ✖ fail: "detection engine api security and spaces enabled find_statuses "after each" hook for "should return a single rule status when a single rule is loaded from a find status with defaults added""
01:59:54               │
01:59:54               └-> "after all" hook
01:59:54             └-> "after all" hook
01:59:54         │
01:59:54         │42 passing (2.0m)
01:59:54         │1 failing
```

So this should fix it to where the afterEach calls try up to 20 times before giving up and then on giving up they move on with the hope a different test doesn't fail.


### Checklist
- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios

tylersmalley committed:
[build] Creates Linux aarch64 archive (#69165)

- Updates Linux Chromium builds to accept architecture argument (defaults to x64) for arm64 support.
  - Example: `python ~/chromium/build_chromium/build.py 312d84c8ce62810976feda0d3457108a6dfff9e6 arm64`
- Updates all Chromium builds to include architecture in filename. 
  - `chromium-312d84c-linux_arm64.zip` _(new)_
  - `chromium-312d84c-linux.zip` > `chromium-312d84c-linux_x64.zip`
- Moves Chromium install from data directory to `x-pack/plugins/reporting/chromium`
- Moves Chromium download cache from `x-pack/plugins/reporting/.chromium` to `.chromium`
- Installs Chromium during build (closes #53664)
- Updates build to be architecture aware (x64 and aarch64)
- Removed Chromium debug logs, they were not helpful and can not be written inside the Kibana root. If we were to keep them, we would need to write to `logging.dist`.

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

thompsongl committed:
Upgrade EUI to v26.3.1 (#70243)

* eui to 26.1.0

* Breadcrumb -> EuiBreadcrumb

* keyCode -> key

* src snapshot updates

* update header variables file location

* keyCode -> key in tests

* x-pack simple snapshot updates

* euitooltip updates

* euioverlaymask updates

* testbed keycode -> key

* fixed header sass mixin

* euidatagrid test workaround

* euioverlapmask mock

* euipaginationbutton markup

* nav snapshot update

* eui to 26.3.0

* flyout z-index fix

* euiflyout snapshot updates

* style snapshot updates

* table actions have machine readable text

* table actions have machine readable text

* core api docs update

* table actions snapshot update

* eui to 26.3.1

* update table actions type

* mock issue links; timeout done

* breadcrumb -> euibreadcrumb

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

alisonelizabeth committed:
address index templates feedback (#71353)

angorayc committed:
[SIEM] fix tooltip of notes (#71342)

* fix tooltip of notes

* fix unit test

* update notes tooltip

* fix unit test

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

ogupte committed:
PR feedback

Commit:
[Logs UI] Limit `extendDatemath` to valid ranges (#71113)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

bmorelli25 committed:
[APM] UI text updates (#71333)

elasticmachine committed:
Merge branch 'master' into apm-70859-transaction-duration-anomaly-detection
